scripts/asr_language_modeling/neural_rescorer/eval_neural_rescorer.py

```diff
@@ -0,0 +1,226 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This script would evaluate a neural language model (Transformer) trained with
+`examples/nlp/language_modeling/transformer_lm.py' as a rescorer for ASR systems.
+Given a trained TransformerLMModel `.nemo` file, this script can be used to re-score the beams obtained from a beam
+search decoder of an ASR model.
+
+USAGE:
+1. Obtain `.tsv` file with beams and their corresponding scores. Scores can be from a regular beam search decoder or
+   in fusion with an N-gram LM scores. For a given beam size `beam_size` and a number of examples
+   for evaluation `num_eval_examples`, it should contain (`beam_size` x `num_eval_examples`) lines of
+   form `beam_candidate_text \t score`. This file can be generated by `scripts/asr_language_modeling/ngram_lm/eval_beamsearch_ngram.py`.
+
+2. Rescore the candidates:
+    python eval_neural_rescorer.py
+        --lm_model=[path to .nemo file of the LM]
+        --beams_file=[path to beams .tsv file]
+        --beam_size=[size of the beams]
+        --eval_manifest=[path to eval manifest .json file]
+        --batch_size=[batch size used for inference on the LM model]
+        --alpha=[the value for the parameter rescorer_alpha]
+        --beta=[the value for the parameter rescorer_beta]
+
+You may find more info on how to use this script at:
+https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/main/asr/asr_language_modeling.html
+
+"""
+
+import contextlib
+import json
+from argparse import ArgumentParser
+
+import editdistance
+import numpy as np
+import pandas as pd
+import torch
+import tqdm
+
+from nemo.collections.nlp.models.language_modeling import TransformerLMModel
+from nemo.utils import logging
+
+
+class BeamSearchDataset(torch.utils.data.Dataset):
+    def __init__(self, data_path, tokenizer, manifest_path, beam_size=128, max_seq_length=256):
+        self.data = pd.read_csv(data_path, delimiter="\t", header=None)
+        self.tokenizer = tokenizer
+        self.ground_truths = []
+        with open(manifest_path, 'r') as f_orig:
+            for line in f_orig:
+                item = json.loads(line)
+                self.ground_truths.append(item['text'])
+        self.beam_size = beam_size
+        self.max_seq_length = max_seq_length
+
+    def __len__(self):
+        return len(self.data)
+
+    def __getitem__(self, idx):
+        text = str(self.data[0][idx])
+        tokens = [self.tokenizer.bos_id] + self.tokenizer.text_to_ids(text) + [self.tokenizer.eos_id]
```

Comment:
Is this the Tokenizer of the Neural LM or the Tokenizer of the Acoustic LM?

Reply:
This is the tokenizer of Neural LM. We do not use the acoustic model in this script directly.

Reply:
Ok makes sense. Docstring should note that I guess.